File: android2_lesson03/src/java/java/com/example/ama/android2_lesson03/ui/search/mvp/SearchOnTheMapPresenter.java
package com.example.ama.android2_lesson03.ui.search.mvp;

import android.util.Log;

import com.example.ama.android2_lesson03.repo.SearchQueryManager;
import com.example.ama.android2_lesson03.repo.base.SearchManager;
import com.example.ama.android2_lesson03.ui.base.BasePresenter;
import com.example.ama.android2_lesson03.ui.search.base.SearchOnTheMapView;
import com.example.ama.android2_lesson03.ui.search.base.SearchPresenter;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;

/**
 * Presenter implementation for {@link SearchOnTheMapView}
 *
 * @param <T> view for work with
 */
public class SearchOnTheMapPresenter<T extends SearchOnTheMapView>
        extends BasePresenter<T>
        implements SearchPresenter<T> {

    private SearchManager queryManager;

    public SearchOnTheMapPresenter() {
        queryManager = new SearchQueryManager();
    }

    @Override
    public void loadSavedState() {
        queryManager.loadSavedState((title, address, position, zoom) -> {
            if (view != null) {
                view.moveMapCamera(position, zoom, false);
                view.showOnInnerMap(title, address, position);
            }
        });
    }

    @Override
    public void saveState(Marker currentMarker) {
        queryManager.saveState(currentMarker);
    }

    @Override
    public void findAddressByQuery(String query) {
        queryManager.getFullLocationName(query, (fullLocationName, latLng, zoom) -> {
            if (view != null) {
                view.moveMapCamera(latLng, zoom, true);
                view.showOnInnerMap(null, fullLocationName, latLng);
            }
        });
    }

    @Override
    public void findAddressByLatLng(LatLng latLng) {
        queryManager.getFullLocationName(latLng, (fullLocationName, latLng1, zoom) -> {
            if (view != null) {
                view.moveMapCamera(latLng1, zoom, true);
                view.showOnInnerMap(null, fullLocationName, latLng1);
            }
        });
    }

    @Override
    public void sendQueryToGMapsApp(Marker currentMarker, LatLng cameraPosition, float zoom) {
        queryManager.prepareUriForGMaps(currentMarker, cameraPosition, zoom, uri -> {
            if (view != null) {
                view.showOnGMapsApp(uri);
            }
        });
    }

    @Override
    public void findMyLocation() {
        queryManager.getMyLocation(new SearchManager.OnLatLngSearchResultCallback() {
            @Override
            public void onLocationFound(LatLng latLng, float zoom) {
                if (view != null) {
                    view.moveMapCamera(latLng, zoom, true);
                }
            }

            @Override
            public void onNotFound(String message) {
                if (view != null) {
                    view.showMessage(message);
                }
            }
        });
    }

    @Override
    public void onMarkerClick(Marker marker) {
        queryManager.prepareSaveMarkerDialog(marker, (title, message, marker1) -> {
            if (view != null) {
                view.showDialog(title, message, marker1);
            }
        });
    }

    @Override
    public void saveMarker(Marker currentMarker, String customName) {
        queryManager.saveMarkerToList(currentMarker, customName, message -> {
            if (view != null) {
                view.showMessage(message);
            }
        });
    }

    @Override
    public void subscribeOnLocationUpdates() {
        queryManager.subscribeOnLocationUpdates(new SearchManager.OnLatLngSearchResultCallback() {
            @Override
            public void onLocationFound(LatLng latLng, float zoom) {
                Log.d("onLocationFound", "location updated " + latLng.toString());
            }

            @Override
            public void onNotFound(String message) {
                if (view != null) {
                    view.showMessage(message);
                }
            }
        });
    }

    @Override
    public void unsubscribeOfLocationUpdates() {
        queryManager.unsubscribeOfLocationUpdates();
    }
}

File: android2_lesson03/src/kotlin/java/com/example/ama/android2_lesson03/ui/search/mvp/SearchOnTheMapPresenter.kt
package com.example.ama.android2_lesson03.ui.search.mvp

import com.example.ama.android2_lesson03.repo.SearchQueryManager
import com.example.ama.android2_lesson03.repo.base.SearchManager
import com.example.ama.android2_lesson03.ui.base.BasePresenter
import com.example.ama.android2_lesson03.ui.search.base.SearchOnTheMapView
import com.example.ama.android2_lesson03.ui.search.base.SearchPresenter
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker

/**
 * Presenter implementation for [SearchOnTheMapView]
 * @param <T> view for work with
 */
class SearchOnTheMapPresenter<T : SearchOnTheMapView> : BasePresenter<T>(), SearchPresenter<T> {

    private val queryManager: SearchManager = SearchQueryManager()

    override fun findAddressByQuery(query: String) {
        queryManager.getFullLocationName(query,
                { fullLocationName, position, zoom ->
                    view?.moveMapCamera(position, zoom, true)
                    view?.showOnInnerMap(null, fullLocationName, position)
                })
    }

    override fun findAddressByLatLng(latLng: LatLng) {
        queryManager.getFullLocationName(latLng,
                { fullLocationName, position, zoom ->
                    view?.moveMapCamera(position, zoom, true)
                    view?.showOnInnerMap(null, fullLocationName, position)
                })
    }

    override fun sendQueryToGMapsApp(currentMarker: Marker?, cameraPosition: LatLng?, zoom: Float?) {
        queryManager.prepareUriForGMaps(currentMarker, cameraPosition, zoom) { uri -> view?.showOnGMapsApp(uri) }
    }

    override fun findMyLocation() {
        queryManager.getMyLocation(
                found = { latLng, zoom -> view?.moveMapCamera(latLng, zoom, true) },
                notFound = { message -> view?.showMessage(message) },
                permissionRequired = { permission, requestCode ->
                    view?.requestPermission(permission, requestCode)
                })
    }

    override fun saveMarker(marker: Marker, customName: String) {
        queryManager.saveMarkerToList(marker, customName)
        { message -> view?.showMessage(message) }
    }

    override fun onMarkerClick(marker: Marker) {
        queryManager.prepareSaveMarkerDialog(marker,
                { dialogTitle, dialogMessage, savingMarker ->
                    view?.showEditDialog(dialogTitle, dialogMessage, savingMarker)
                })
    }

    override fun saveState(currentMarker: Marker?) {
        queryManager.saveState(currentMarker)
    }

    override fun loadSavedState() {
        queryManager.loadSavedState { markerTitle, address, position, zoom ->
            view?.moveMapCamera(position, zoom, false)
            view?.showOnInnerMap(markerTitle, address, position)
        }
    }

    override fun subscrineOnLocationUpdates() {
        queryManager.subscribeOnLocationUpdates(
                found = { latLng, zoom -> },
                notFound = { message -> view?.showMessage(message) },
                permissionRequired = { permission, requestCode -> view?.requestPermission(permission, requestCode) })
    }

    override fun unsubscribeOfLocationUpdates() {
        queryManager.unsubscribeOfLocationUpdates()
    }
}
